sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1050 @@
+
+|===
+
+==== Explicit Graph Building API
+
+When using the explicit graph building API to construct a graph, nodes and
+edges are captured as follows.
+
+Table 3. Explicit Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| In the explicit graph building API nodes are created by the user invoking
+methods on a modifiable graph. Each node represent either a command-group
+function, empty operation, or device memory allocation/free.
+
+| Edge
+| In the explicit graph building API edges are defined by the user. This is
+either through buffer accessors, the `make_edge()` function, or by passing
+dependent nodes on creation of a new node.
+|===
+
+==== Queue Recording API
+
+When using the record & replay API to construct a graph by recording a queue,
+nodes and edges are captured as follows.
+
+Table 4. Recorded Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| Nodes in a queue recorded graph represent each of the command group
+submissions of the program. Each submission encompasses either one or both of
+a.) some data movement, b.) a single asynchronous kernel launch. Nodes cannot
+define forward edges, only backwards (i.e. kernels can only create dependencies
+on things that have already happened). This means that transparently a node can
+depend on a previously recorded graph (sub-graph), which works by creating edges
+to the individual nodes in the old graph. Explicit memory operations without
+kernels, such as a memory copy, are still classed as nodes under this
+definition, as the {explicit-memory-ops}[SYCL 2020 specification states] that
+these can be seen as specialized kernels executing on the device.
+
+| Edge
+| An edge in a queue recorded graph represents a data dependency between two
+nodes. Data dependencies can naturally be expressed in user code through buffer
+accessors. USM pointers also convey data dependencies, however offsets into
+system allocations (`malloc`/`new`) are not supported.
+|===
+
+=== API Modifications
+
+[source, c++]
+----
+namespace sycl {
+namespace ext::oneapi::experimental {
+
+// State of a queue, returned by info::queue::state
+enum class queue_state {
+  executing,
+  recording
+};
+
+class node {};
+
+// State of a graph
+enum class graph_state {
+  modifiable,
+  executable
+};
+
+// New object representing graph
+template<graph_state State = graph_state::modifiable>
+class command_graph {};
+
+template<>
+class command_graph<graph_state::modifiable> {
+public:
+  command_graph(const property_list &propList = {});
+  command_graph<graph_state::executable> finalize(const context &syclContext) const;
+
+  node add(const std::vector<node>& dep = {});
+
+  template<typename T>
+  node add(T cgf, const std::vector<node>& dep = {});
+
+  node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& dep = {});
+  node add_free(void *data, const std::vector<node>& dep = {});
+
+  void make_edge(node sender, node receiver);
```

Comment:
Agree that this terminology isn't the most readable, think `src` and `dest` would be the clearest

Reply:
Okay, I'll think about renaming `sender` and `receiver`. 😊

There is the option to pass an arbitrary number of predecessors to a node via the `add` function when it's constructed. I just reworked how dependencies are passed using a `property` instead of a `std::vector`. Please have a look.

Not sure if we necessarily need the `make_edge` overload for multiple senders, you can always call it in a loop. `node` handles are intended to be organized by a user. That could be in any container, not necessarily convertible to a `std::vector`.